src/types/mongo/stardust.rs

```diff
@@ -0,0 +1,668 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::str::FromStr;
+
+use anyhow::{
+    anyhow,
+    bail,
+};
+use mongodb::bson::{
+    from_bson,
+    to_bson,
+    Bson,
+    DateTime,
+    Document,
+};
+
+use super::{
+    BsonExt,
+    DocExt,
+};
+use crate::stardust::{
+    address::{
+        Address,
+        AliasAddress,
+        Ed25519Address,
+        NftAddress,
+    },
+    input::{
+        Input,
+        TreasuryInput,
+        UtxoInput,
+    },
+    milestone::MilestoneIndex,
+    output::{
+        AliasId,
+        AliasOutput,
+        BasicOutput,
+        FeatureBlocks,
+        FoundryOutput,
+        NativeTokens,
+        NftId,
+        NftOutput,
+        Output,
+        TokenTag,
+        TreasuryOutput,
+        UnlockConditions,
+    },
+    parent::Parents,
+    payload::{
+        milestone::{
+            MilestoneEssence,
+            MilestoneId,
+            MilestonePayload,
+        },
+        receipt::{
+            MigratedFundsEntry,
+            ReceiptPayload,
+            TailTransactionHash,
+        },
+        tagged_data::TaggedDataPayload,
+        transaction::{
+            RegularTransactionEssence,
+            TransactionEssence,
+            TransactionId,
+            TransactionPayload,
+        },
+        treasury_transaction::TreasuryTransactionPayload,
+        Payload,
+    },
+    signature::{
+        Ed25519Signature,
+        Signature,
+    },
+    unlock_block::{
+        AliasUnlockBlock,
+        NftUnlockBlock,
+        ReferenceUnlockBlock,
+        SignatureUnlockBlock,
+        UnlockBlock,
+        UnlockBlocks,
+    },
+    Message,
+    MessageBuilder,
+    MessageId,
+};
+
+pub fn message_to_bson(message: &Message) -> Bson {
+    let mut doc = Document::new();
+    doc.insert("protocol_version", message.protocol_version() as i32);
+    doc.insert(
+        "parents",
+        message.parents().iter().map(|p| p.to_string()).collect::<Vec<_>>(),
+    );
+    doc.insert("payload", message.payload().map(payload_to_bson));
+    doc.insert("nonce", message.nonce().to_string());
+    Bson::Document(doc)
+}
```

Comment:
There are several problems with storing the message directly.
1. Bson cannot store u64, at all, ever. If we have u64 data, it will try to convert to an i64 and fail if it is out of range. This is only important for the nonce as far as I can tell, but it needs to be explicitly de/serialized as a string.
Now, if that is the only u64 that can possibly be out of range, then we can simply manually handle the top level and use the serde impl for the payload.
2. In order to create consistent indexes on the data, we need to add fields to the stored document in strategic places. For instance, several of the `index-able` fields are inside lists of blocks in stardust, but some, like the address, are fundamentally the same index as cpt2. Therefore, the simplest solution to index both is to create a field in the same place on both documents. This has to be at the output level conceptually, so it can't just be at the root. Similarly, timestamps must be converted to Bson `Dates` to be useful, versus directly serializing them as i64.
3. The `bee` serde impls use externally tagged enums, meaning that all levels are effectively two levels (i.e. `payload.essence.outputs.address` vs `payload.data.essence.data.outputs.data.address.data`) which makes matching and pipelines much more difficult. If we're going to use their impls, I think it would be in our best interest to change that to internal tags.

I don't think it really makes sense to change the protocol to support Bson. Ideally this data could be stored anywhere, but I think it's reasonable to expect that some internal conversions must take place. I think we could definitely reduce this and the effort needed for each protocol release by serializing the payloads and adding fields to the resulting document as needed, but there are good reasons why I did it this way for the time being.